server/src/DBTablesLastInfo.cc

```diff
@@ -0,0 +1,483 @@
+		ItemGroupStream itemGroupStream(*itemGrpItr);
+		LastInfoDef lastInfoDef;
+
+		itemGroupStream >> lastInfoDef.id;
+		itemGroupStream >> lastInfoDef.dataType;
+		itemGroupStream >> lastInfoDef.value;
+		itemGroupStream >> lastInfoDef.serverId;
+
+		lastInfoDefList.push_back(lastInfoDef);
+	}
+
+	return HTERR_OK;
+}
+
+static string makeIdListCondition(const LastInfoIdList &idList)
+{
+	string condition;
+	const ColumnDef &colId = COLUMN_DEF_LAST_INFO[IDX_LAST_INFO_ID];
+	condition = StringUtils::sprintf("%s in (", colId.columnName);
+	LastInfoIdListConstIterator it = idList.begin();
+	while (true) {
+		const int id = *it;
+		condition += StringUtils::sprintf("%d", id);
+		++it;
+		if (it == idList.end())
+			break;
+		condition += ",";
+	}
+	condition += ")";
+	return condition;
+}
+
+static string makeServerIdCondition(string serverId)
+{
+	return StringUtils::sprintf(
+		 "%s=%" FMT_LAST_INFO_SERVER_ID,
+		 COLUMN_DEF_LAST_INFO[IDX_LAST_INFO_SERVER_ID].columnName,
+		 serverId.c_str());
+}
+static string makeConditionForDelete(const LastInfoIdList &idList,
+				     const OperationPrivilege &privilege)
+{
+	string condition = makeIdListCondition(idList);
+
+	return condition;
+}
+
+HatoholError DBTablesLastInfo::deleteLastInfoList(const LastInfoIdList &idList,
+                                                  const OperationPrivilege &privilege)
+{
+	HatoholError err = checkPrivilegeForDelete(privilege, idList);
+	if (err != HTERR_OK)
+		return err;
+
+	if (idList.empty()) {
+		MLPL_WARN("idList is empty.\n");
+		return HTERR_INVALID_PARAMETER;
+	}
+
+	struct TrxProc : public DBAgent::TransactionProc {
+		DBAgent::DeleteArg arg;
+		uint64_t numAffectedRows;
+
+		TrxProc (void)
+		: arg(tableProfileLastInfo),
+		  numAffectedRows(0)
+		{
+		}
+
+		void operator ()(DBAgent &dbAgent) override
+		{
+			dbAgent.deleteRows(arg);
+			numAffectedRows = dbAgent.getNumberOfAffectedRows();
+		}
+	} trx;
+	trx.arg.condition = makeConditionForDelete(idList, privilege);
+	getDBAgent().runTransaction(trx);
+
+	// Check the result
+	if (trx.numAffectedRows != idList.size()) {
+		MLPL_ERR("affectedRows: %" PRIu64 ", idList.size(): %zd\n",
+		         trx.numAffectedRows, idList.size());
+		return HTERR_DELETE_INCOMPLETE;
+	}
+
+	return HTERR_OK;
+}
+
+// ---------------------------------------------------------------------------
+// Protected methods
+DBTables::SetupInfo &DBTablesLastInfo::getSetupInfo(void)
+{
+	static const TableSetupInfo DB_TABLE_INFO[] = {
+	{
+		&tableProfileLastInfo,
+	},
+	};
+
+	static SetupInfo setupInfo = {
+		DB_TABLES_ID_LAST_INFO,
+		LAST_INFO_DB_VERSION,
+		ARRAY_SIZE(DB_TABLE_INFO),
+		DB_TABLE_INFO,
+		updateDB,
+	};
+	return setupInfo;
+};
+
+// ---------------------------------------------------------------------------
+HatoholError DBTablesLastInfo::checkPrivilegeForAdd(
+  const OperationPrivilege &privilege, const LastInfoDef &actionDef)
+{
+	UserIdType userId = privilege.getUserId();
```

Comment:
I like to use 'const' for the read only variable.
Of course, there is no big problem if it doesn't